src/NuGet.Core/Microsoft.Build.NuGetSdkResolver/GlobalJsonReader.cs

```diff
@@ -0,0 +1,142 @@
+                || !TryGetPathOfFileAbove(GlobalJsonFileName, projectDirectory.FullName, out var globalJsonPath))
+            {
+                return null;
+            }
+
+            var contents = File.ReadAllText(globalJsonPath);
+
+            // Look ahead in the contents to see if there is an msbuild-sdks section.  Deserializing the file requires us to load
+            // Newtonsoft.Json which is 500 KB while a global.json is usually ~100 bytes of text.
+            if (contents.IndexOf(MSBuildSdksPropertyName, StringComparison.Ordinal) == -1)
+            {
+                return null;
+            }
+
+            try
+            {
+                return Deserialize(contents);
+            }
+            catch (Exception e)
+            {
+                // Failed to parse "{0}". {1}
+                var message = string.Format(CultureInfo.CurrentCulture, Strings.FailedToParseGlobalJson, globalJsonPath, e.Message);
+                context.Logger.LogMessage(message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Deserializes a global.json and returns the MSBuild SDK versions
+        /// </summary>
+        private static Dictionary<string, string> Deserialize(string value)
+        {
+            return JsonConvert.DeserializeObject<GlobalJsonFile>(value).MSBuildSdks;
+        }
+
+        private sealed class GlobalJsonFile
+        {
+            [JsonProperty(MSBuildSdksPropertyName)]
+            public Dictionary<string, string> MSBuildSdks { get; set; }
+        }
+
+        /// <summary>
+        /// Searches for a file based on the specified starting directory.
+        /// </summary>
+        /// <param name="file">The file to search for.</param>
+        /// <param name="startingDirectory">An optional directory to start the search in.  The default location is the directory
+        /// of the file containing the property function.</param>
+        /// <returns>The full path of the file if it is found, otherwise an empty string.</returns>
+        private static string GetPathOfFileAbove(string file, string startingDirectory)
+        {
+            // Search for a directory that contains that file
+            var directoryName = GetDirectoryNameOfFileAbove(startingDirectory, file);
+
+            return string.IsNullOrEmpty(directoryName) ? string.Empty : NormalizePath(Path.Combine(directoryName, file));
+        }
+
+        private static bool TryGetPathOfFileAbove(string file, string startingDirectory, out string fullPath)
+        {
+            fullPath = GetPathOfFileAbove(file, startingDirectory);
+
+            return fullPath != string.Empty;
+        }
+
+        /// <summary>
+        /// Locate a file in either the directory specified or a location in the
+        /// directory structure above that directory.
+        /// </summary>
+        private static string GetDirectoryNameOfFileAbove(string startingDirectory, string fileName)
```

Comment:
Is there a NuGet API that can do this?  I know they have code to search for NuGet.config